### ISSUE TITLE
SurfaceTool tutorial/doc about index() wrong or confusing

### ISSUE THREAD

**Issue by pseidemann** [godotengine/godot-docs#6386]
**Your Godot version:**

stable

**Issue description:**

on the page "Using the SurfaceTool", the index() method is briefly mentioned.
what I understand from this page is that index() removes duplicate vertices. how it knows which vertices are duplicate is not mentioned and unknown to me (I suspect identical positions? what about rounding errors?).

"[...] calling index() which shrinks the vertex array to remove duplicate vertices."

there is this short snippet:
```gdscript
# Creates a quad from four corner vertices.
# Add_index does not need to be called before add_vertex.
st.add_index(0)
st.add_index(1)
st.add_index(2)

st.add_index(1)
st.add_index(3)
st.add_index(2)

# Alternatively:
st.index()
```

this implies that if I add 4 vertices and then call index(), I will get a quad. from the documentation of index() it's not clear that this is what will happen. indeed I cannot make index() work to create a quad for me without using add_index().

I suspect the documentation is wrong here or quite confusing. I don't know which information is correct to fix it myself. what is index() really doing?

the documentation in the class for this method is also not clear about it:
"Shrinks the vertex array by creating an index array. This can improve performance by avoiding vertex reuse."

glad for any insights!

**URL to the documentation page:**

https://docs.godotengine.org/en/stable/tutorials/3d/procedural_geometry/surfacetool.html

https://docs.godotengine.org/en/stable/classes/class_surfacetool.html#class-surfacetool-method-index

**Comment by clayjohn** [comment]
More detail could indeed be helpful. ``index()`` essentially runs through the list of vertices and notes which are duplicates (duplicate vertices are vertices with the same position and attribute info (normal, UV, etc.), then it shrinks the vertex array to only contain unique vertices, and generates an index array that maintains the original mesh structure. 

To turn a list of vertices into a quad you would need 6 vertices. ``index()`` would then collapse that into 4 vertices and an index array of length 6.

**Comment by pseidemann** [comment]
> To turn a list of vertices into a quad you would need 6 vertices. `index()` would then collapse that into 4 vertices and an index array of length 6.

this is actually not true. the snippet from the doc is pretty bad, but if you add 4 vertices and then add 6 indices (add_index(), not index()) all in the right order you get a mesh. I got this working after some trial and error.
see also this for reference: https://gamedev.stackexchange.com/a/10741

**Comment by clayjohn** [comment]
> > To turn a list of vertices into a quad you would need 6 vertices. `index()` would then collapse that into 4 vertices and an index array of length 6.
> 
> this is actually not true. the snippet from the doc is pretty bad, but if you add 4 vertices and then add 6 indices (add_index(), not index()) all in the right order you get a quad. I got this working after some trial and error. see also this for reference: https://gamedev.stackexchange.com/a/10741
> 
> so replacing 6 calls of add_index() with one call of index() is what the tutorial suggests but this seems to be entirely wrong, hence the bug report.

Yes, thats why I said you would need 6 vertices to reproduce the same behaviour, not 4

**Comment by pseidemann** [comment]
@clayjohn,
gotcha :)

ok this is interesting. my natural next question would be how index() is able to form a quad (with setting indices) from 6 vertices where 2 are duplicates, when it fails to do so for 4 unique ones?

from your post:
"[...] and generates an index array that maintains the original mesh structure.":
it's not clear to me what you mean by mesh structure (we are building one right now so there is no structure [yet] [?]).

speculating what index() does:

- remember duplicate vertices
- assume that the first 3 vertices are one triangle and the next 3 are another triangle (and so forth?)
- now it knows how to merge the two triangles by removing the two duplicates and setting the indices in a way that the 4 vertices form the quad

this opens some new questions for me:

how does this behave when I create two disjoint quads?

also maybe even more interesting, what happens if I create multiple quads and some of them are overlapping (by 1 or 2 vertices)?

once we gathered the info, I'm happy to create a PR to improve the tutorial+doc about index() :)

**Comment by clayjohn** [comment]
> ok this is interesting. my natural next question would be how index() is able to form a quad (with setting indices) from 6 vertices where 2 are duplicates, when it fails to do so for 4 unique ones?

When you call ``begin()`` you tell the SurfaceTool what kind of geometry you are creating. In the doc, we use ``PRIMITIVE_TRIANGLES`` which produces a triangle out of every three vertices. If you give it only 4 vertices it will error when you try to commit the mesh. 

> "[...] and generates an index array that maintains the original mesh structure.":
it's not clear to me what you mean by mesh structure (we are building one right now so there is no structure [yet] [?]).

This means the resulting mesh will look the same whether you call ``index()`` or not. In other words, ``index()`` will not move vertices around or change vertex attributes. 

> speculating what index() does:
>- remember duplicate vertices
>- assume that the first 3 vertices are one triangle and the next 3 are another triangle (and so forth?)
>- now it knows how to merge the two triangles by removing the two duplicates and setting the indices in a way that the 4 vertices form the quad

No assumptions are needed. It knows how to treat the vertices because you instruct it what primitive to use when calling ``begin()``. Further ``index()`` doesn't need to know anything about the underlying vertices it just assigns them a number, the vertex ordering is up to you. 

Maybe a more detailed example will help. Lets say we generate a quad like so:
```
var st = SurfaceTool.new()

st.begin(Mesh.PRIMITIVE_TRIANGLES)

st.add_vertex(Vector3(-1, -1, 0)) # 0
st.add_vertex(Vector3(-1, 1, 0)) # 1
st.add_vertex(Vector3(1, 1, 0)) # 2

st.add_vertex(Vector3(-1, -1, 0)) #3 - duplicate of vertex-0
st.add_vertex(Vector3(1, 1, 0)) # 4 - duplicate of vertex-2
st.add_vertex(Vector3(1, -1, 0)) # 5

# Commit to a mesh.
var mesh = st.commit()
```

When calling ``index()`` it will identify that vertices 3 and 4 are duplicates of vertices 0 and 2 respectively. ``index()`` runs through the list of vertices and maintains a list of unique positions. If the vertex positions is not in the list, we add it to the list and add an index with the position of the vertex in the unique positions list. If the vertex position is in the list, we add an index with the position of the vertex in the list. At the end, the list of indices will have the same number of elements as there were vertices in the original vertex array. But the new vertex array will have fewer elements. 

So let's look at the process that ``index()`` takes (roughly speaking):

```
# Vertex 0: Vector3(-1, -1, 0)
# not in unique vertex array, so add.
# unique vertex array = [Vector3(-1, -1, 0)]
add_index(0); 

# Vertex 1: Vector3(-1, 1, 0)
# not in unique vertex array, so add.
# unique vertex array = [Vector3(-1, -1, 0), Vector3(-1, 1, 0)]
add_index(1); 

# Vertex 2: Vector3(1, 1, 0)
# not in unique vertex array, so add.
# unique vertex array = [Vector3(-1, -1, 0), Vector3(-1, 1, 0), Vector3(1, 1, 0)]
add_index(2); 

# Vertex 3: Vector3(-1, -1, 0)
# found in unique vertex array at position 0, so do not add.
# unique vertex array =  [Vector3(-1, -1, 0), Vector3(-1, 1, 0), Vector3(1, 1, 0)]
add_index(0); 

# Vertex 4: Vector3(1, 1, 0)
# found in unique vertex array at position 2, so do not add.
# unique vertex array =  [Vector3(-1, -1, 0), Vector3(-1, 1, 0), Vector3(1, 1, 0)]
add_index(2); 

# Vertex 5: Vector3(-1, -1, 0)
# not in unique vertex array, so add.
# unique vertex array = [Vector3(-1, -1, 0), Vector3(-1, 1, 0), Vector3(1, 1, 0), Vector3(1, -1, 0)]
add_index(3); 

# At the end of index() we have
vertex_array = [Vector3(-1, -1, 0), Vector3(-1, 1, 0), Vector3(1, 1, 0), Vector3(1, -1, 0)]
index_array = [0, 1, 2, 0, 2, 3]
```

**Comment by pseidemann** [comment]
in the end your vertex_array has 4 vertices.
isn't this a contradiction to your first paragraph: 
"If you give it only 4 vertices it will error when you try to commit the mesh."?

otherwise the algorithm of index() is now roughly clear to me. I think what is not so easy is to deduce how this algorithm can be utilized for different use cases besides just drawing a single quad.

**Comment by clayjohn** [comment]
> in the end your vertex_array has 4 vertices.
> isn't this a contradiction to your first paragraph:
>"If you give it only 4 vertices it will error when you try to commit the mesh."?

No its not. In my example I gave the SurfaceTool 6 vertices and then called ``index()``. The important thing here is telling the SurfaceTool what surface to form. If I give it only 4 vertices and I am using PRIMITIVE_TRIANGLES, there isn't enough information to form a quad. If I give it 6 vertices and I am using PRIMITIVE_TRIANGLES then there is enough information to form a quad (assuming the vertices are arranged to form a quad).

> otherwise the algorithm of index() is now roughly clear to me. I think what is not so easy is to deduce how this algorithm can be utilized for different use cases besides just drawing a single quad (referencing to my last two questions in the previous post here).

It works the exact same way. The index array has the exact same structure that the original vertex array did. If you create two disjoint quads (12 vertices) you will end up with 8 vertices in your vertex array and 12 indices. 

Remember, the index array follows the structure of the vertex array. It isn't doing anything clever to the surface you have formed. It copies the exact structure you used in the vertex array to form the surface but does so with a list of indices instead of storing every single vertex individually. If your vertex array forms a bunch of disjoint triangles, so will the index array. If your vertex array creates a nice surface with lots of shared triangles, then the index array will create that same surface

**Comment by pseidemann** [comment]
> If your vertex array creates a nice surface with lots of shared triangles, then the index array will create that same surface

I tried now to use the index() technique with adding 6 vertices for one quad.

I did a test with a 2x2 grid of 4 quads.
I would expect the following numbers:
number of vertices: 9
number of edges: 16

however, when I measure these two numbers with MeshDataTool, I get these numbers:
number of vertices: 16
number of edges: 20

now comes the twist... when I render the same 2x2 quad grid with the 4 vertices technique and add_index(), I get the the expected numbers (9 vertices, 16 edges).

is this expected? from what I understood is that index() should remove these redundant vertices/edges but that doesn't seem to happen and index() is actually "less performant" in that regard.

am I missing something?
